snow/engine/common/timer.go

```diff
@@ -3,12 +3,87 @@
 
 package common
 
-import "time"
+import (
+	"sync"
+	"time"
+)
 
-// Timer describes the standard interface for specifying a timeout
-type Timer interface {
+// PreemptionSignal signals when to preempt the timeouts of the timeout handler.
+type PreemptionSignal struct {
+	activateOnce sync.Once
+	initOnce     sync.Once
+	signal       chan struct{}
+}
+
+func (ps *PreemptionSignal) init() {
+	ps.signal = make(chan struct{})
+}
+
+// Listen returns a read-only channel that is closed when Preempt() is invoked.
+func (ps *PreemptionSignal) Listen() <-chan struct{} {
+	ps.initOnce.Do(ps.init)
+	return ps.signal
+}
+
+// Preempt causes any past and future call of Listen return a closed channel.
+func (ps *PreemptionSignal) Preempt() {
+	ps.initOnce.Do(ps.init)
+	ps.activateOnce.Do(func() {
+		close(ps.signal)
+	})
+}
+
+type timeoutHandler struct {
+	newTimer         func(time.Duration) (<-chan time.Time, func() bool)
+	timeouts         chan struct{}
+	onTimeout        func()
+	preemptionSignal <-chan struct{}
+}
```

Comment:
I made it this way because I didn't want to change the original code pattern, and to make as minimal changes as possible. 

I now re-implemented it. I hope it reads better and is easier to reason about.